app/lib/storage/vault.ts

```diff
@@ -0,0 +1,121 @@
+
+    // This is the asynchronous part of the constructor.
+    this.ready = new Promise(async resolve => {
+      // Ensure the LevelDB directory exists
+      await Utils.ensurePath(dbPath)
+
+      // Create and set a new encryption options object with everything set to
+      // default but the password attribute, which is user defined.
+      this.cryptOptions = new Encryption.Options(
+        { password } as Encryption.IProtoOptions
+      )
+
+      // Create and set the underlying Level (levelup + leveldown) database.
+      this.db = level(dbPath, {
+        keyEncoding: "binary",
+        valueEncoding: "binary"
+      })
+
+      // This causes the `this.ready` promise to resolve.
+      resolve(true)
+    })
+  }
+
+  /**
+   * Create or overwrite a value entry for a certain key.
+   * @param {string} key
+   * @param value
+   * @returns {Promise<boolean>}
+   */
+  public async put(key: string, value: any): Promise<boolean> {
+    // Keys are hashed for privacy and to allow arbitrary key size
+    const hashedKey = Encryption.keyHash(key)
+    // Dehydrate/deflate and serialize the object to convert it into a typed
+    // JSON that can be rehydrated/inflated afterwards without losing types.
+    const serializedValue = Buffer.from(JSON.stringify(hydra.dehydrate(value)))
+    // Encrypt the serialized object.
+    const encryptedValue =
+      Encryption.encrypt(serializedValue, this.cryptOptions)
+    // Put the value into the database and wait for write to complete.
+    await new Promise(resolve => {
+      this.db.put(hashedKey, encryptedValue, resolve)
+    })
+
+    return true
+  }
+
+  /**
+   * Get the value associated with a certain key.
+   * @param {string} key
+   * @returns {Promise<any>}
+   */
+  public async get(key: string): Promise<any> {
+    // Hash the key for privacy and to allow arbitrary key size.
+    const hashedKey = Encryption.keyHash(key)
+    // Retrieve the encrypted value.
+    const encryptedValue = await new Promise((resolve, reject) => {
+      this.db.get(hashedKey, (error, value) => {
+        if (error) {
+          reject(error)
+        } else {
+          resolve(value)
+        }
+      })
+    })
+    // Ensure encryptedValue is a Buffer
+    if (!encryptedValue) {
+      console.log(encryptedValue, typeof encryptedValue)
+      throw new TypeError("encryptedValue is not a Buffer")
+    }
+    // Decrypt the encrypted value from the database.
+    const serializedValue =
+      Encryption.decrypt(encryptedValue as Buffer, this.cryptOptions)
+
+    // Hydrate/inflate the decrypted value to recover the original object with
+    // types and everything.
+    return hydra.hydrate(serializedValue)
+  }
+
+  /**
+   * Close the vault, the LevelDB database and the underlying storage, freeing
+   *  up any locks and file descriptors.
+   * @returns {Promise<any>}
+   */
+  public async close(): Promise<any> {
+    this.db.close()
+  }
+
+}
```

Comment:
Here the same, I'm totally ok with it, although I'm more fan on this alternative implementation:

1. Define the interface I'm going to use in the rest of the program to access the key-value store engine, for example (the name of the interface is irrelevant, is just the one that first came to mind):

```typescript
interface IStore {
  put(key: string, value: any): Promise<boolean>;
  get(key: string): Promise<any>;
  close(): void
}
```

2. Define a concrete implementation for that interface, in our case it would be a leveldb-backed implementation:

```typescript
class Vault implements IStore {
  constructor(private name: string, private db: any, private cryptOptions)
  // implementation of methods
  // ....
}
```
where `db` is already an active connection to leveldb

3. Define a factory function that is responsible of creating an instance of the concrete implementation for `IStore`:

```typescript
function createLevelDbVault(name: string, password: string, path?: string): Promise<Vault> {
  return new Promise(async resolve => {
    // ...
    // This causes the `this.ready` promise to resolve.
    resolve(new LevelDbVault(name, db, cryptOptions))
  })
}
```
The reason to use a factory function instead of putting that logic inside the class constructor is to avoid having a component with too much responsibility, if for some reason the way the db connection is created is enriched, all we need to do is create a new factory function to accommodate the change. 

The advantage of this model is that now, whenever I'm testing a component of the application that needs an `IStore` value, I can pass one more suitable for testing, e.g. one that keeps the data in-memory.

Reply:
I like your way of thinking...

Same as before, I'll give a try to your suggested pattern. I find it indeed to be quite idiomatic. Forgive me if I got carried away too much by the object-oriented nature of TypeScript :rofl: